### ISSUE TITLE
Error after checking my spellbook while fighting a monster

### ISSUE THREAD

**Issue by Nillx** [butter-cookie-kitkat/ButterQuestTracker#53]
I feel like an error magnet at this point :D Got this error after checking which spells im missing. Using this addon: https://www.wowinterface.com/downloads/info25031-WhatsTraining.html

```
1x [ADDON_ACTION_BLOCKED] AddOn 'ButterQuestTracker' tried to call the protected function 'SpellButton2:Show()'.
!BugGrabber\BugGrabber.lua:519: in function <!BugGrabber\BugGrabber.lua:519>
[C]: in function `Show'
FrameXML\SpellBookFrame.lua:185: in function `SpellBookFrame_UpdateSpells'
FrameXML\SpellBookFrame.lua:769: in function `SpellBook_UpdatePlayerTab'
FrameXML\SpellBookFrame.lua:14: in function `tabUpdate'
FrameXML\SpellBookFrame.lua:160: in function <FrameXML\SpellBookFrame.lua:112>
[C]: in function `SpellBookFrame_Update'
FrameXML\SpellBookFrame.lua:97: in function <FrameXML\SpellBookFrame.lua:96>
[C]: in function `Show'
FrameXML\UIParent.lua:2078: in function `SetUIPanel'
FrameXML\UIParent.lua:1923: in function `ShowUIPanel'
FrameXML\UIParent.lua:1784: in function <FrameXML\UIParent.lua:1780>
[C]: in function `SetAttribute'
FrameXML\UIParent.lua:2526: in function `ShowUIPanel'
FrameXML\SpellBookFrame.lua:46: in function `ToggleSpellBook'
[string "TOGGLESPELLBOOK"]:1: in function <[string "TOGGLESPELLBOOK"]:1>

Locals:
InCombatSkipped
```

Got this one after checking the default blizzard spellbook

```
1x [ADDON_ACTION_BLOCKED] AddOn 'ButterQuestTracker' tried to call the protected function 'SpellButton2:Enable()'.
!BugGrabber\BugGrabber.lua:519: in function <!BugGrabber\BugGrabber.lua:519>
[C]: in function `Enable'
FrameXML\SpellBookFrame.lua:493: in function `SpellButton_UpdateButton'
FrameXML\SpellBookFrame.lua:271: in function `SpellButton_OnEvent'
[string "*:OnEvent"]:1: in function <[string "*:OnEvent"]:1>

Locals:
InCombatSkipped
```

**Comment by cecilia-sanare** [comment]
Well that's certainly different! 

Looking into it now. Haha

**Comment by cecilia-sanare** [comment]
So I literally just spent the past hour debugging this and I'm going to attempt to explain what the issue was in an attempt to retain my sanity.

However first I thought I'd let you know that this issue should be resolved in [1.9.0-rc.4](https://www.curseforge.com/wow/addons/butter-quest-tracker/files/2801981).

Now onto the ramblings of a mad man!

**Basically the specific line of code causing the problem was here...**

https://github.com/butter-cookie-kitkat/ButterQuestTracker/blob/4d623facd955dc2eeb092df471a5e2436adad4e4/TrackerHelper/Modules/Font.lua#L45

**More specifically the issue was with the default values I was passing up here...**

https://github.com/butter-cookie-kitkat/ButterQuestTracker/blob/4d623facd955dc2eeb092df471a5e2436adad4e4/TrackerHelper/Modules/Font.lua#L21-L22

~~These values contain rgb values, but they also contain a whole bunch of functions.
I'm guessing it had something to do with how WoWs UIFrameCache reuses frames in conjunction with this fact that was causing the problem.~~

After digging into it further I'm convinced it's because of the rgb assignments (`value.r = value.r or 0.0`) I had in that function. Basically I was updating the globally shared color variables Blizzard provides `NORMAL_FONT_COLOR` and `HIGHLIGHT_FONT_COLOR`.

When the spellbook tried to access these modified variables it ended up causing problems.

**Upon updating the output of `_normalizeColor`..**

https://github.com/butter-cookie-kitkat/ButterQuestTracker/blob/8beae4eb3d2b0b4295db55752e98b91076dc178b/TrackerHelper/Modules/Base.lua#L37-L48

**To instead only return the rgba values...**

https://github.com/butter-cookie-kitkat/ButterQuestTracker/blob/4d623facd955dc2eeb092df471a5e2436adad4e4/TrackerHelper/Modules/Base.lua#L45-L58

~~The issue magically went away.~~
The issue went away because we're returning a new object / table.

Now you may be asking yourself, how in the world did you figure that out?
Simply by commenting out segments of the code and narrowing down where the issue was.

**Comment by Nillx** [comment]
Pretty weirdly sounding bug! Wish i was able to provide some help but sadly my coding skills are extremely basic. 

Ill install rc4 and ill let you know if i find more issues.
